k8s/src/main/scala/io/buoyant/k8s/EndpointsNamer.scala

```diff
@@ -105,270 +108,283 @@ class SingleNsNamer(
+                val newMap = service.portMappings
+                logEvent.deletion(oldMap -- newMap.keySet)
+                newMap
+              case (oldMap, v1.ServiceError(error)) =>
+                log.warning(
+                  "k8s ns %s service %s watch error %s",
+                  nsName, serviceName, error
+                )
+                oldMap
             }
-          case None =>
-            service.port(portName).map {
-              case Some(addr) =>
-                log.debug("k8s ns %s service %s port %s found + %s", nsName, serviceName, portName, residual.show)
-                Activity.Ok(NameTree.Leaf(Name.Bound(addr, idPrefix ++ id, residual)))
-              case None =>
-                log.debug("k8s ns %s service %s port %s missing", nsName, serviceName, portName)
-                Activity.Ok(NameTree.Neg)
-            }
-        }
-        Activity(state)
     }
-  }
 
   /**
-   * For a given port number, apply the port mapping of the service.  The target port of the port
-   * mapping may be a named port and the named port may or may not exist.  The outer Var[Option]
-   * of the return type tracks whether the port exists and the inner Var[Addr] tracks the actual
-   * endpoints if the port does exist.
+   * Watch the numbered-port remappings for the service named `serviceName`
+   * in the namespace named `nsName`.
+   * @param nsName the name of the Kubernetes namespace.
+   * @param serviceName the name of the Kubernetes service.
+   * @return an `Activity` containing a `Map[Int, String]` representing the
+   *         port number to port name mappings
+   * @note that the corresponding `Activity` instances are cached so we don't
+   *       create multiple watches on the same Kubernetes objects – meaning,
+   *       if you look up the port remappings in a given (nsName, serviceName)
+   *       pair multiple times, you will always get back the same `Activity`,
+   *       which is created the first time that pair is looked up.
    */
-  private[this] def lookupNumberedPort(
-    serviceCache: ServiceCache,
-    svc: SvcCache,
+  private[this] def numberedPortRemappings(
+    nsName: String,
     serviceName: String,
-    portNumber: Int
-  ): Var[Option[Var[Addr]]] =
-    serviceCache.getPortMapping(serviceName, portNumber).flatMap {
-      case Some(targetPort) =>
-        targetPort.flatMap { target =>
-          // target may be an int (port number) or string (port name)
-          Try(target.toInt).toOption match {
-            case Some(targetPortNumber) =>
-              // target port is a number and therefore exists
-              Var(Some(svc.port(targetPortNumber)))
-            case None =>
-              // target port is a name and may or may not exist
-              svc.port(target)
-          }
-        }
-      case None =>
-        Var(None)
+    labelSelector: Option[String]
+  ) = portRemappingsMemo((nsName, serviceName, labelSelector))
```

Comment:
My tendency is to annotate all defs, but skip it for single line vals. Not sure how consistently that's applied in the codebase though.